File: P12/conVolSecuencial.cpp
//
// Created by isa on 18/01/19.
//
#include <iostream>
#include <cmath>
#include <random>
#include <ctime>
#include <iomanip>

struct MatrizImagen {
    static const int LIM_MAX = 20;
    static const int LIM_INF = -20;
    int **matriz;

    MatrizImagen() {
        matriz = new int*[10000];
        for(int i = 0; i < 10000; i++)
            matriz[i] = new int[10000];
    }

    std::array<std::array<int, 3>, 3> celdasAdyacentes(int x, int y) const {
        static std::array<std::array<int, 3>, 3> adyacentes;

        for(int i = 0; i < 3; i++)
            for(int j = 0; j < 3; j++)
                adyacentes[i][j] = 0;

        for(int i = -1; i <= 1; i++)
            for(int j = -1; j <= 1; j++)
                if((x + i) >= 0 && (x + i) < 10000 && (y + j) >= 0 && (y + j) < 10000)
                    adyacentes[i + 1][j + 1] = matriz[x][y];

        return adyacentes;
    }

    void imprimirMatriz(){
        for(int i = 0; i < 10000; i++)
        {
            for(int j = 0; j < 10000; j++)
                std::cout << std::setw(3) << matriz[i][j] << " ";
        std::cout << std::endl;
        }
    }

    ~MatrizImagen(){
        for(int i = 0; i < 10000; i++)
            delete[] matriz[i];

        delete[] matriz;
    }
};

void inicializarMatriz(MatrizImagen& matrizImagen){
    srand (time(NULL));
    for(int i = 0; i < 10000; i++)
        for(int j = 0; j < 10000; j++)
            matrizImagen.matriz[i][j] = MatrizImagen::LIM_INF + rand() % (MatrizImagen::LIM_MAX - MatrizImagen::LIM_INF);
}

int aplicarFiltro(const int filtro[3][3], const std::array<std::array<int, 3>, 3>& subMatriz){
    int resultado = 0;

    for(int i = 0; i < 3; i++)
        for(int j = 0; j < 3; j++)
            resultado += filtro[i][j] * subMatriz[i][j];

    if(resultado > MatrizImagen::LIM_MAX)
        resultado = MatrizImagen::LIM_MAX;

    if(resultado <  MatrizImagen::LIM_INF)
        resultado = MatrizImagen::LIM_INF;

    return resultado;
}

MatrizImagen convolucionarMatriz(const MatrizImagen& matrizInicial){
    static const int sobel[][3] = {{-1, 0, 1}, {-2, 0, 2}, {-1, 0, 1}};
    static const int sharpen[][3] = {{1, -2, 1}, {-2, 5, 2}, {1, -2, 1}};
    int op;
    MatrizImagen matrizConvolucionada;

    for(int i = 0; i < 10000; i++)
        for(int j = 0; j < 10000; j++)
        {
            std::cout << "¿Qué filtro desea aplicar?" << std::endl;
            std::cout << "1.Sobel" << std::endl;
            std::cout << "2.Sharpen" << std::endl;
            std::cin >> op;

            std::array<std::array<int, 3>, 3> subMatriz = matrizInicial.celdasAdyacentes(i, j);

            switch (op){
                case 1:
                    matrizConvolucionada.matriz[i][j] = aplicarFiltro(sobel, subMatriz);
                    break;
                case 2:
                    matrizConvolucionada.matriz[i][j] = aplicarFiltro(sharpen, subMatriz);
                    break;
            }

        }
    return matrizConvolucionada;
}

int main(){
    MatrizImagen matrizImagen;
    inicializarMatriz(matrizImagen);

    convolucionarMatriz(matrizImagen);
    return 0;
}

File: P12/conVolSecuencial.java
import java.util.Random;

/**
 * @author isa
 *
 */
public class conVolSecuencial {
	static final int LIM_MAX = 20;
    static final int LIM_INF = -20;
    
    static Random r = new Random();
    
    /**
     * @param matriz
     */
    static void inicializarMatriz(int[][] matriz){
    	for(int i = 0; i < 10000; i++)
    		for(int j = 0; j < 10000; j++)
    			matriz[i][j] = -20 + r.nextInt(40);
    		
    }
    
    /**
     * @param matriz
     */
    static void imprimirMatriz(int[][] matriz) {
    	for(int i = 0; i < 10000; i++)
    		for(int j = 0; j < 10000; j++)
    			System.out.println(matriz[i][j]);
    }
    
    /**
     * @param matIn
     * @param matOut
     * @param fila
     * @param columna
     */
    static void celdasAdyacentes(int[][] matIn, int[][] matOut, int fila, int columna){	
    	for (int i = 0; i < matOut.length; i++)
    		for (int j = 0; j < matOut.length; j++) 
    			matOut[i][j] = 0;
    	
    	for(int i = -1; i <= 1; i++)
    		for(int j = -1; j <= 1; j++) {
    			if((fila + i) >= 0 && (fila + i) < 10000 && (columna + j) >= 0 && (columna + j) < 10000)
    				matOut[i + 1][j + 1] = matIn[fila + i][columna + j];
    		}		
    }
    
    /**
     * @param matrizAdyacentes
     * @param filtro
     * @return
     */
    static int sumarAdyacentes(int[][] matrizAdyacentes, int[][] filtro) {
    	int resultado = 0;
    	for(int i = 0; i < matrizAdyacentes.length; i++)
			for(int j = 0; j < matrizAdyacentes.length; j++)
				resultado += matrizAdyacentes[i][j] * filtro[i][j];
    	
    	if(resultado > LIM_MAX)
    		resultado = LIM_MAX;
    	if(resultado < LIM_INF)
    		resultado = LIM_INF;
    	
    	return resultado;
				
    }
    
    /**
     * @param matInicial
     * @param matOut
     * @param filtro
     */
    static void aplicarFiltro(int[][] matInicial, int[][] matOut, int[][] filtro) {
    	int[][] matrizAdyacentes = new int[3][3];
    	
    	for(int i = 0; i < matInicial.length; i++)
    		for(int j = 0; j < matInicial.length; j++) {
    			celdasAdyacentes(matInicial, matrizAdyacentes, i, j);
    			matOut[i][j] = sumarAdyacentes(matrizAdyacentes, filtro);
    			
    		}
    }
    
    /**
     * @param matInicial
     * @param matConvolucionadaSobel
     * @param matConvolucionadaSharpen
     */
    static void convolucionarMatriz(int[][] matInicial, int[][] matConvolucionadaSobel, int[][] matConvolucionadaSharpen) {
    	int[][] sobel = {{-1, 0, 1}, {-2, 0, 2}, {-1, 0, 1}};
    	int[][] sharpen = {{1, -2, 1}, {-2, 5, -2}, {1, -2, 1}};
    	
    	aplicarFiltro(matInicial, matConvolucionadaSobel, sobel);
    	aplicarFiltro(matInicial, matConvolucionadaSharpen, sharpen);
    	
    }
    
    /**
     * @param args
     */
    public static void main(String[] args) {
    	int[][] matriz = new int[10000][10000];
        int[][] matConvolucionadaSobel = new int[10000][10000];
        int[][] matConvolucionadaSharpen = new int[10000][10000];
        
        inicializarMatriz(matriz);
        
        double tInicio;
        double tFin;
        tInicio = System.currentTimeMillis();
        convolucionarMatriz(matriz, matConvolucionadaSobel, matConvolucionadaSharpen);
        tFin = System.currentTimeMillis();
        System.out.println(tFin - tInicio);

	}
}
